Enable more clippy rules

UnknownPlatypus opened the pull request:
> Maybe one PR per lint?

I went for a single PR because it was easier to manage and because a lot of lint either require few changes or trigger a cascading rule.

Review commit per commit is required, I'd like to get your take on some lint rules (I'm not convinced by a few ones).
The last commit ignore some rules directly, I've added a comment on each one with the rationale.

Fixes #221

LilyFirefly commented on src/render/filters.rs:
```diff
@@ -541,7 +541,7 @@ impl ResolveFilter for YesnoFilter {
         };
 
-        Ok(Some(result.to_string().into_content()))
+        Ok(Some((*result).to_string().into_content()))
```
I'm not sure we need to fix this one - it [only affects old compiler versions](https://rust-lang.github.io/rust-clippy/master/index.html?search=inefficient_to_string) which I don't think we care to support.

UnknownPlatypus replied:
oh, the documentation changed on master. This was not mentioned for rust [1.91](https://rust-lang.github.io/rust-clippy/rust-1.91.0/index.html?search=inefficient_to_string#/inefficient_to_string)

I'm ignoring this one then

LilyFirefly commented on src/render/tags.rs:
```diff
@@ -851,12 +851,12 @@ fn build_kwargs<'py>(
     for (key, value) in kwargs {
         let value = value.resolve(py, template, context, ResolveFailures::Raise)?;
-        _kwargs.set_item(key, value)?;
+        new_kwargs.set_item(key, value)?;
```
```suggestion
        py_kwargs.set_item(key, value)?;
```

LilyFirefly commented on src/render/filters.rs:
```diff
@@ -83,9 +83,9 @@ impl ResolveFilter for AddSlashesFilter {
-                        raw.replace(r"\", r"\\")
-                            .replace("\"", "\\\"")
-                            .replace("'", r"\'"),
+                        raw.replace('\\', r"\\")
```
Sad!

LilyFirefly commented:
> > I don't like changing context to ctx everywhere - I think it's significantly worse than what we had.
> 
> The flagged issue is about `context` vs `content` because there is only a one letter difference which arguably could cause issues rather easily. Do you have another suggestion on how to help differentiate these ? I'm fine disabling this lint otherwise

I'd rather rename `content` then. Maybe `variable` or `node` or `part` or `element`? Different names might make sense in different places.

UnknownPlatypus commented:
> I'd rather rename content then. Maybe variable or node or part or element? Different names might make sense in different places.

I'll defer this to a separate PR, this is quite large changes but it makes more sense, a lot of things are named content whil being rather different objects. I've disabled `similar_names` for now

I've rebased the MR and fixed the various issues.
i've also added `match_same_arms` which I missed somehow

LilyFirefly commented on src/render/tags.rs:
```diff
@@ -613,11 +609,9 @@ impl Evaluate for IfCondition {
                     (Some(Content::Bool(left)), Some(Content::Py(right))) => {
                         !right.is(PyBool::new(py, left).as_any())
                     }
-                    (Some(Content::Bool(_)), _) => true,
```
Why is this removal valid?

LilyFirefly commented on src/render/tags.rs:
```diff
@@ -592,7 +589,6 @@ impl Evaluate for IfCondition {
                     (Some(Content::Py(obj)), None) | (None, Some(Content::Py(obj))) => {
                         obj.is(PyNone::get(py).as_any())
                     }
-                    (Some(Content::Bool(_)), None) => false,
```
Oh, nevermind! I see it.
